File: platform/pylib/scout_common/observability/tracing/metadata.py
# platform/pylib/scout_common/observability/tracing/metadata.py
"""gRPC metadata helpers for request ID propagation."""

from typing import Optional, Tuple

import grpc

from .request_id import RequestID, REQUEST_ID_HEADER
from .context import get_request_id, set_request_id


def inject_request_id(
    metadata: Optional[list] = None,
    request_id: Optional[RequestID] = None
) -> list:
    """Inject request ID into gRPC outgoing metadata.
    
    Used by gRPC clients to propagate request IDs to downstream services.
    
    Args:
        metadata: Existing metadata list (or None for new)
        request_id: Request ID to inject (or None to use from context)
        
    Returns:
        Metadata list with request ID injected
    """
    if metadata is None:
        metadata = []
    
    if request_id is None:
        request_id = get_request_id()
    
    if request_id is None:
        return metadata
    
    # Remove existing request ID if present
    metadata = [
        (k, v) for k, v in metadata 
        if k.lower() != REQUEST_ID_HEADER.lower()
    ]
    
    # Add request ID
    metadata.append((REQUEST_ID_HEADER, str(request_id)))
    return metadata


def extract_request_id(context: grpc.ServicerContext) -> Optional[RequestID]:
    """Extract request ID from gRPC incoming metadata.
    
    Used by gRPC servers to receive request IDs from upstream services.
    
    Args:
        context: gRPC servicer context
        
    Returns:
        Request ID if found, None otherwise
    """
    metadata = dict(context.invocation_metadata())
    
    # Try case-insensitive lookup
    for key in [REQUEST_ID_HEADER, REQUEST_ID_HEADER.lower()]:
        value = metadata.get(key)
        if value:
            return RequestID.parse(value)
    
    return None


def extract_or_generate_request_id(
    context: grpc.ServicerContext
) -> RequestID:
    """Extract request ID from metadata or generate a new one.
    
    Also sets it in the context for downstream use.
    
    Args:
        context: gRPC servicer context
        
    Returns:
        Request ID (extracted or newly generated)
    """
    request_id = extract_request_id(context)
    if request_id is None:
        request_id = RequestID.generate()
    
    set_request_id(request_id)
    return request_id


def propagate_request_id(
    metadata: Optional[list] = None
) -> list:
    """Propagate request ID from context to outgoing metadata.
    
    Useful for client interceptors to automatically propagate request IDs.
    
    Args:
        metadata: Existing metadata list (or None for new)
        
    Returns:
        Metadata list with request ID propagated
    """
    request_id = get_request_id()
    if request_id is None:
        request_id = RequestID.generate()
        set_request_id(request_id)
    
    return inject_request_id(metadata, request_id)

File: platform/pkg/observability/tracing/metadata.go
// platform/pkg/observability/tracing/metadata.go
package tracing

import (
	"context"

	"google.golang.org/grpc/metadata"
)

// InjectRequestID injects the request ID into gRPC outgoing metadata.
// This is used by gRPC clients to propagate request IDs to downstream services.
func InjectRequestID(ctx context.Context, requestID RequestID) context.Context {
	if requestID == "" {
		return ctx
	}

	md, ok := metadata.FromOutgoingContext(ctx)
	if !ok {
		md = metadata.New(nil)
	} else {
		// Clone to avoid modifying shared metadata
		md = md.Copy()
	}

	md.Set(RequestIDHeader, requestID.String())
	return metadata.NewOutgoingContext(ctx, md)
}

// ExtractRequestID extracts the request ID from gRPC incoming metadata.
// This is used by gRPC servers to receive request IDs from upstream services.
// Returns empty RequestID if not found.
func ExtractRequestID(ctx context.Context) RequestID {
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return ""
	}

	values := md.Get(RequestIDHeader)
	if len(values) == 0 {
		return ""
	}

	return ParseRequestID(values[0])
}

// ExtractOrGenerateRequestID extracts the request ID from gRPC metadata,
// or generates a new one if not found, and attaches it to the context.
func ExtractOrGenerateRequestID(ctx context.Context) (context.Context, RequestID) {
	rid := ExtractRequestID(ctx)
	if rid == "" {
		rid = NewRequestID()
	}
	return WithRequestID(ctx, rid), rid
}

// PropagateRequestID extracts the request ID from incoming metadata
// and injects it into outgoing metadata for downstream calls.
// This is useful in middleware/interceptors.
func PropagateRequestID(ctx context.Context) context.Context {
	rid := GetRequestID(ctx)
	if rid == "" {
		rid = ExtractRequestID(ctx)
		if rid == "" {
			rid = NewRequestID()
		}
		ctx = WithRequestID(ctx, rid)
	}
	return InjectRequestID(ctx, rid)
}